### PR TITLE
Manage dnf-query.py as a Cookbook File.

### LINKED ISSUE #14: Test Kitchen fails to deploy `libraries/dnf-query.py`
Test kitchen appears to be omitting the python file from the libraries directory.

```
 * dnf_package[git] action install

           ================================================================================
           Error executing action `install` on resource 'dnf_package[git]'
           ================================================================================

           Errno::ENOENT
           -------------
           No such file or directory - /tmp/kitchen/cache/cookbooks/dnf/libraries/dnf-query.py

           Cookbook Trace:
           ---------------
           /tmp/kitchen/cache/cookbooks/dnf/libraries/provider_dnf_package.rb:112:in `available_version'
           /tmp/kitchen/cache/cookbooks/dnf/libraries/provider_dnf_package.rb:77:in `block in load_current_resource'
           /tmp/kitchen/cache/cookbooks/dnf/libraries/provider_dnf_package.rb:75:in `each'
           /tmp/kitchen/cache/cookbooks/dnf/libraries/provider_dnf_package.rb:75:in `load_current_resource'

           Resource Declaration:
           ---------------------
           # In /tmp/kitchen/cache/cookbooks/omnibus-chef-arm/recipes/git_config.rb

            20:   dnf_package 'git'
            21: else

           Compiled Resource:
           ------------------
           # Declared in /tmp/kitchen/cache/cookbooks/omnibus-chef-arm/recipes/git_config.rb:20:in `from_file'

           dnf_package("git") do
             provider Chef::Provider::Package::Dnf
             action [:install]
             retries 0
             retry_delay 2
             default_guard_interpreter :default
             package_name "git"
             declared_type :dnf_package
             cookbook_name "omnibus-chef-arm"
             recipe_name "git_config"
           end


       Running handlers:
       [2016-01-28T18:21:30-05:00] ERROR: Running exception handlers
       Running handlers complete
       [2016-01-28T18:21:30-05:00] ERROR: Exception handlers complete
       Chef Client failed. 0 resources updated in 22 seconds
       [2016-01-28T18:21:30-05:00] FATAL: Stacktrace dumped to /tmp/kitchen/cache/chef-stacktrace.out
       [2016-01-28T18:21:30-05:00] ERROR: dnf_package[git] (omnibus-chef-arm::git_config line 20) had an error: Errno::ENOENT: No such file or directory - /tmp/kitchen/cache/cookbooks/dnf/libraries/dnf-query.py
       [2016-01-28T18:21:32-05:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)
```

While this is likely bug in test-kitchen, I tried identifying what causes the exclusion in kitchen, and haven't had much luck. I suspect it will likely be easier to resolve this issue in the cookbook instead.

### PR DESCRIPTION
**Pull request by rhass**
There is a bug in the current version of Test Kitchen which causes the `dnf-query.py` script to be delete during the cleanup. (See test-kitchen/test-kitchen#929 for details.) This changes the way the file is dropped by using the Chef `cookbook_file` resource, and places the file in a common path.

Fixes #14

[Review on Reviewable](https://reviewable.io/reviews/pantheon-cookbooks/dnf/15)

### CODE REVIEW

File: libraries/provider_dnf_package.rb [hunk 1]
```diff
@@ -102,7 +102,14 @@ def installed_version(package_name)
     end
 
     def dnf_query_helper
-      ::File.join(::File.dirname(__FILE__), 'dnf-query.py')
+      query_helper = ::File.join(Chef::Config['file_cache_path'], 'dnf-query.py')
+
+      f = Chef::Resource::CookbookFile.new(query_helper, run_context)
+      f.mode(0755)
+      f.cookbook('dnf')
```
Comment (rhass) [inline libraries/provider_dnf_package.rb]: Hard-coding `f.cookbook('dnf')` is super ugly, but `self.cookbook_name` returns the value of a calling cookbook, so I am not what object I can run `cookbook_name` against to automagically resolve to the correct value.

**Comment by fujin** [comment]
I took a quick look at this and can't get it to converge in TK or tests to pass with chefspec, seeing:

```
Failures:

  1) Chef::Provider::Package::Dnf load_current_resource when querying for package state should set the version number when there is an available package
     Failure/Error: expect(@provider.available_version('nc6')).to eq('0:1.0-30.fc22')
     NameError:
       uninitialized constant Chef::Resource::CookbookFile
     # ./libraries/provider_dnf_package.rb:107:in `dnf_query_helper'
     # ./libraries/provider_dnf_package.rb:120:in `available_version'
     # ./test/spec/libraries/provider_dnf_package_spec.rb:92:in `block (3 levels) in <top (required)>'

  2) Chef::Provider::Package::Dnf load_current_resource when querying for package state should not set the version number when there is not an available package
     Failure/Error: expect(@provider.available_version('nc6')).to be_nil
     NameError:
       uninitialized constant Chef::Resource::CookbookFile
     # ./libraries/provider_dnf_package.rb:107:in `dnf_query_helper'
     # ./libraries/provider_dnf_package.rb:120:in `available_version'
     # ./test/spec/libraries/provider_dnf_package_spec.rb:98:in `block (3 levels) in <top (required)>'
```

And in TK:
```
-----> Chef Omnibus installation detected (11.16.4)
D      [SSH] kitchen@192.168.99.100<{:user_known_hosts_file=>"/dev/null", :paranoid=>false, :port=>32770, :compression=>true, :compression_level=>6, :keepalive=>true, :keepalive_interval=>60, :timeout=>15, :password=>"kitchen", :user=>"kitchen"}> (sh -c '

sudo_rm="sudo -E rm"
dirs="/tmp/kitchen/clients /tmp/kitchen/cookbooks /tmp/kitchen/data /tmp/kitchen/data_bags /tmp/kitchen/encrypted_data_bag_secret /tmp/kitchen/environments /tmp/kitchen/roles"
root_path="/tmp/kitchen"

$sudo_rm -rf $dirs
mkdir -p $root_path
')
       Transferring files to <install-fedora-22-chef-11164>
D      Uploaded /var/folders/_v/p7wj2kc95nsdv25y367m32980000gn/T/install-fedora-22-chef-11164-sandbox-20160202-37310-18v3hrq/cookbooks/dnf/README.md (3558 bytes)
D      Uploaded /var/folders/_v/p7wj2kc95nsdv25y367m32980000gn/T/install-fedora-22-chef-11164-sandbox-20160202-37310-18v3hrq/dna.json (49 bytes)
D      Uploaded /var/folders/_v/p7wj2kc95nsdv25y367m32980000gn/T/install-fedora-22-chef-11164-sandbox-20160202-37310-18v3hrq/solo.rb (628 bytes)
D      Transfer complete
D      [SSH] kitchen@192.168.99.100<{:user_known_hosts_file=>"/dev/null", :paranoid=>false, :port=>32770, :compression=>true, :compression_level=>6, :keepalive=>true, :keepalive_interval=>60, :timeout=>15, :password=>"kitchen", :user=>"kitchen"}> (sh -c '

sudo -E /opt/chef/bin/chef-solo --config /tmp/kitchen/solo.rb --log_level debug --force-formatter --no-color --json-attributes /tmp/kitchen/dna.json
')
       [2016-02-02T19:17:05+00:00] INFO: Forking chef instance to converge...
       [2016-02-02T19:17:05+00:00] DEBUG: Fork successful. Waiting for new chef pid: 834
       [2016-02-02T19:17:05+00:00] DEBUG: Forked instance now converging
       [2016-02-02T19:17:05+00:00] WARN:
       * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
       SSL validation of HTTPS requests is disabled. HTTPS connections are still
       encrypted, but chef is not able to detect forged replies or man in the middle
       attacks.

       To fix this issue add an entry like this to your configuration file:

       ```
         # Verify all HTTPS connections (recommended)
         ssl_verify_mode :verify_peer

         # OR, Verify only connections to chef-server
         verify_api_cert true
       ```

       To check your SSL configuration, or troubleshoot errors, you can use the
       `knife ssl check` command like so:

       ```
         knife ssl check -c /tmp/kitchen/solo.rb
       ```

       * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *

       Starting Chef Client, version 11.16.4
       [2016-02-02T19:17:05+00:00] INFO: *** Chef 11.16.4 ***
       [2016-02-02T19:17:05+00:00] INFO: Chef-client pid: 834
       [2016-02-02T19:17:05+00:00] DEBUG: Chef-client request_id: 4e973a03-f970-47bf-aa2c-d8c4d656fab5
       [2016-02-02T19:17:05+00:00] WARN: unable to detect ipaddress
       [2016-02-02T19:17:05+00:00] WARN: unable to detect macaddress
       [2016-02-02T19:17:05+00:00] DEBUG: Building node object for install-fedora-22-chef-11164
       [2016-02-02T19:17:05+00:00] DEBUG: Extracting run list from JSON attributes provided on command line
       [2016-02-02T19:17:05+00:00] INFO: Setting the run_list to ["recipe[dnf_package_install_test]"] from CLI options
       [2016-02-02T19:17:05+00:00] DEBUG: Applying attributes from json file
       [2016-02-02T19:17:05+00:00] DEBUG: Platform is fedora version 22
       [2016-02-02T19:17:05+00:00] INFO: Run List is [recipe[dnf_package_install_test]]
       [2016-02-02T19:17:05+00:00] INFO: Run List expands to [dnf_package_install_test]
       [2016-02-02T19:17:05+00:00] INFO: Starting Chef Run for install-fedora-22-chef-11164
       [2016-02-02T19:17:05+00:00] INFO: Running start handlers
       [2016-02-02T19:17:05+00:00] INFO: Start handlers complete.
       [2016-02-02T19:17:05+00:00] DEBUG: No chefignore file found at /tmp/kitchen/chefignore no files will be ignored
       Compiling Cookbooks...
       [2016-02-02T19:17:05+00:00] DEBUG: Re-raising exception: Chef::Exceptions::CookbookNotFound - Cookbook dnf_package_install_test not found. If you're loading dnf_package_install_test from another cookbook, make sure you configure the dependency in your metadata
       /opt/chef/embedded/lib/ruby/gems/1.9.1/gems/chef-11.16.4/lib/chef/cookbook/cookbook_collection.rb:38:in `block in initialize'
```

Looks like for whatever reason, the cooks and stuff aren't making it into the VM during the berks transfer. Adding the `require`'s for the other Chef classes gets a little out of hand as well, plus it looks like the `Chef::Config['file_cache_path']` doesn't exist during chefspec run time, has to be created prior to make spec, or managed.

I'm wondering if there's not a better way we can do this, like using Poise so we can do LWRP-style cookbook_file instead of invoking the resource directly onto the hwrp's context.

Do the specs work for you? How about TK? @rhass. Thanks!

**Comment by rhass-r7** [comment]
Sorry about the fail :whale:. I was hacking this on a flight home, and may have omitted testing in kitchen due to the slowness of downloading containers over inflight wifi. I will validate this and get back to you (also having weird issues with kitchen-docker at the moment).

I think your original intentions of using a HWRP instead of a custom resource or a poised gem was to make it easier to port over to mainline chef code. Nevertheless, I think a rewrite is not far from what may need to be done; and as much as I love working with poise, with a potentially low-level resource used for installing packages I think making it as vanilla as possible is probably the best approach to ensure people don't have dependency collisions with other poised cookbooks later down the line.

**Comment by rhass** [comment]
I seem to be having ssh key issues with the Fedora containers. The SSH keys generated by kitchen-docker 2.3.0 do not seem to get dropped into the container instance. Is there anything I need to add to my .kitchen.local.yml for these docker containers?

**Comment by lamont-granquist** [comment]
https://github.com/chef/chef/pull/4640 should have fixed this problem.

**Comment by rhass** [comment]
@lamont-granquist Cool! Thanks! :smiley:

**Comment by rhass** [comment]
DNF has been merged into chef-client.

**Pull request closed without merging**